File: src/main/java/systems/conduit/main/ConduitConfiguration.java
package systems.conduit.main;

import lombok.Getter;
import lombok.NoArgsConstructor;
import systems.conduit.main.core.plugin.config.Configuration;
import systems.conduit.main.core.plugin.config.annotation.ConfigFile;

/**
 * Main configuration file for Conduit.
 *
 * @author Innectic
 * @since 12/30/2019
 */
@ConfigFile(name = "conduit", type = "json", defaultFile = "conduit.json")
@NoArgsConstructor
public class ConduitConfiguration extends Configuration {

    @Getter
    @NoArgsConstructor
    private class MySQLConfiguration {
        private String host;
        private int port;
        private String username;
        private String password;
        private String database;
    }

    @Getter
    @NoArgsConstructor
    private class RedisConfiguration {
        private String host;
        private int port;
        private String password;
        private int db;
    }

    @Getter
    @NoArgsConstructor
    private class DatastoreConfiguration {
        private MySQLConfiguration mysql;
        private RedisConfiguration redis;
    }

    @Getter private DatastoreConfiguration datastores;
}

File: src/main/kotlin/systems/conduit/core/ConduitConfiguration.kt
package systems.conduit.core

import systems.conduit.core.plugin.config.Configuration
import systems.conduit.core.plugin.config.annotation.ConfigFile

/**
 * Main configuration file for Conduit.
 *
 * @author Innectic
 * @since 12/30/2019
 */
@ConfigFile(name = "conduit", type = "json", defaultFile = "conduit.json")
class ConduitConfiguration: Configuration() {

    inner class MySQLConfiguration {
        var host: String? = null
            private set
        var port = 0
            private set
        var username: String? = null
            private set
        var password: String? = null
            private set
        var database: String? = null
            private set
    }

    inner class RedisConfiguration {
        var host: String? = null
            private set
        var port = 0
            private set
        var password: String? = null
            private set
        var db = 0
            private set
    }

    inner class DatastoreConfiguration {
        var mysql: MySQLConfiguration? = null
            private set
        var redis: RedisConfiguration? = null
            private set
    }

    var datastores: DatastoreConfiguration? = null
        private set
}
